plugin/mmdb/mmdb.go

```diff
@@ -0,0 +1,100 @@
+// Package mmdb implements a max mind database plugin.
+package mmdb
+
+import (
+	"context"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+	"github.com/oschwald/geoip2-golang"
+)
+
+// Mmdb is a plugin that add location data to the request context
+// from a max mind geoIP2 database, which can be later consumed by other middlewares.
+type Mmdb struct {
+	Next        plugin.Handler
+	databases   []mmdb
+	labelValues map[string]string
+}
+
+type mmdb struct {
+	*geoip2.Reader
+	// provides defines the schemas that can be obtained by querying this database, by using
+	// bitwise operations.
+	provides int
+}
+
+const (
+	anonymousIP = 1 << iota
+	asn
+	city
+	connectionType
+	country
+	domain
+	enterprise
+	isp
+)
+
+// ServeDNS implements the plugin.Handler interface.
+func (m Mmdb) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	return plugin.NextOrFailure(pluginName, m.Next, ctx, w, r)
+}
+
+const labelPrefix = "plugin/" + pluginName + "/"
+
+// Metadata implements the metadata.Provider Interface in the metadata plugin, and is used to store
+// the data associated with the source IP of every request.
+func (m Mmdb) Metadata(ctx context.Context, state request.Request) context.Context {
+	srcIP := net.ParseIP(state.IP())
+	if srcIP == nil {
+		// This should be unreachable code.
+		// TODO(snebel29): Print a debug error without causing a logging flood?
+		// TODO(snebel29): Is there any Prometheus metric that we should provide?
+		return ctx
+	}
+
+	for _, db := range m.databases {
+		// Enterprise schema is a superset of City schema which in turn is a superset of Country schema
+		// therefore we just query the schema that has the most data available.
+		switch {
+		case db.provides&enterprise == 1:
+			data, err := db.Enterprise(srcIP)
+			if err != nil {
+				// TODO(snebel29): Print a debug error without causing a logging flood?
+				break
```

Comment:
probably need a way to set an error in a label?

Reply:
oh, interesting. I was naively thinking that consumers would simply look for the keys they are interested and simply handle scenarios when is available and not.

Do you think providing the error in a label would be handy for consumers? do you have an specific label convention in mind?